### ISSUE TITLE
How does Kubeflow use data stored on EFS, S3 and EBS?

### ISSUE THREAD

**Issue by psulowsk** [awslabs/kubeflow-manifests#300]
Hello,
I dont't know how to attach "Question" label so I used another one.

I would like to find out a bit more how Kubeflow uses input data stored on different storages - S3, EFS and EBS. 

EFS seems quite understandable to me - as it is a file system, different Tfjobs and different containers can use the same data on EFS multiple times. 

How does it work if the data is stored on EBS and in S3 bucket? Do Kubeflow retrieve S3 data every time Tfjob is strarted? Are they then temporarily stored, e.g. on EBS? How data is distributed if I use data stored on EBS (as EBS can be attached only to one node)?

Additionally, do you recommend (or dissuade) the use of any storage as an input data source (e.g. S3) for machine learning worklads in Kubeflow?

I would be grateful for your explanation. Thank you!

**Comment by wenjinsitu** [comment]
Currently we support deployment option for RDS and S3 (combined or one of these):
https://awslabs.github.io/kubeflow-manifests/docs/deployment/rds-s3/guide/

Also EFS or FSx as add-on:
https://awslabs.github.io/kubeflow-manifests/docs/deployment/add-ons/storage/efs/guide/

**Comment by surajkota** [comment]
Hi @psulowsk, thanks for your question and apologies for late response.

EBS, EFS and FSx for lustre are different types of storage systems(file/block) which can be attached to EC2 instances.

Kubernetes has the concept of [PersistentVolume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#introduction) which is an abstraction over the storage type and its provisioning. PVs can be mounted as volumes to pods. Persistent volumes are dynamically provisioned using [storage classes](https://kubernetes.io/docs/concepts/storage/storage-classes/) and each StorageClass has a provisioner that determines what volume plugin is used for provisioning PVs. EBS, EFS and FSx are different types of PVs, each of which have a volume plugin which abstracts the provisioning of that type of volume. 

S3 on the other hand is not a file system. It is an object storage service where data is stored as objects in a flat plane (with no hierarchy). It has been designed as a Key-Value store unlike the above storage systems in which all data is organised hierarchically in a top-down network of folders. So, as you have already noticed there is no provisioner to use S3 buckets as volumes.

Coming to Kubeflow, Kubeflow as a platform does not offer anything specific to PVs and relys on the above K8s concepts i.e. using PVs for storing data(datasets, models) but note that some components do offer integration with cloud storage like S3. How to store and use data is upto the user and the ML framework. e.g. 
- when using Training operator, a TfJob expects a `PodSpec` for the worker pod where you can [specify the volumes which should be mounted to the pod](https://github.com/kubeflow/training-operator/blob/master/examples/tensorflow/mnist_with_summaries/tf_job_mnist.yaml#L24-L26). You can pre-cache the dataset from S3 to a volume and reuse it in training job or you can download it at runtime as part of your training script
- Capabilities like data distribution, sharding, batching etc. will be provided by the frameworks you are using e.g. https://pytorch.org/tutorials/intermediate/dist_tuto.html#communication-backends and Kubeflow makes it easy to run these jobs in a K8s native fashion, e.g. https://github.com/kubeflow/training-operator/blob/master/examples/pytorch/mnist/v1/pytorch_job_mnist_gloo.yaml
- For creating an inference service using KServe you can specify the model path via `storageUri` which can be a PV or a cloud path like S3 and the storage initializer will take care of fetching the model and mounting it to the serving container
- Another thing to look up is S3 plugin for the ML frameworks you are using. Some frameworks offer S3 plugin to accelerate download from S3. e.g. [S3 plugin for Pytorch](https://pytorch.org/data/main/generated/torchdata.datapipes.iter.S3FileLoader.html) and [S3 plugin for tensorflow](https://github.com/tensorflow/io/blob/2e50c7a33ba21f651397bb5261b45aa187420092/tests/test_s3.py)

> do you recommend (or dissuade) the use of any storage as an input data source (e.g. S3) for machine learning workloads in Kubeflow?

You can absolutely use S3 as input data source for your ML workload and this is where FSx shines. If your workloads require a fast, POSIX-compliant file system access to your S3 buckets, then you can use FSx for Lustre to link your S3 buckets to a file system and keep data synchronized between the file system and S3 in both directions. You can read more about it on [this blog post](https://aws.amazon.com/blogs/aws/enhanced-amazon-s3-integration-for-amazon-fsx-for-lustre/) and find an example on how to use FSx file system with Kubeflow [here](https://awslabs.github.io/kubeflow-manifests/docs/deployment/add-ons/storage/fsx-for-lustre/guide/).

Hope this helps. Let us know if you have more questions

**Comment by surajkota** [comment]
Closing since there has been no activity in 60 days
